ParadiseSS13/Paradise
# Only give bots HUDs when player-controlled.

## What Does This PR Do

Fixes part of #19150 (spiderbots are not addressed in this PR). This PR addresses GC failures caused by references to simple bots in the global HUD trackers.

Simple bots were more regularly player-controlled in the past. Consequently, the existing implementation always did two things:

1. Added the simple bot to the list of atoms that should have visible HUDs when viewing e.g. as an admin or ghost.
2. Added all the available diagnostic HUDs to the list of HUDs that the simple bot _itself_ should see.

Regardless of whether or not the bot is player controlled, the second reference is never cleared, leading to a GC failure of the simple bot.

```
[2022-09-27T22:22:29] Starting up. Round ID is 108
[2022-09-27T22:22:29] -------------------------
[2022-09-27T22:26:46] Beginning search for references to a /mob/living/simple_animal/bot/secbot/beepsky.
[2022-09-27T22:26:46] Finished searching globals
[2022-09-27T22:27:26] Finished searching atoms
[2022-09-27T22:27:26] Found /mob/living/simple_animal/bot/secbot/beepsky [0x3000006] in list World -> /datum/atom_hud/data/human/security/advanced [0x210000d1] -> hudusers (list).
[2022-09-27T22:27:26] Found /mob/living/simple_animal/bot/secbot/beepsky [0x3000006] in list World -> /datum/atom_hud/data/diagnostic/basic [0x210000d4] -> hudusers (list).
[2022-09-27T22:27:26] Found /mob/living/simple_animal/bot/secbot/beepsky [0x3000006] in list World -> /datum/atom_hud/data/diagnostic/advanced [0x210000d5] -> hudusers (list).
[2022-09-27T22:27:28] Finished searching datums
[2022-09-27T22:27:28] Finished searching clients
[2022-09-27T22:27:28] Completed search for references to a /mob/living/simple_animal/bot/secbot/beepsky.
[2022-09-27T22:27:28] GC: -- [0x2100000d] | /mob/living/simple_animal/bot/secbot/beepsky was unable to be GC'd --
```

This PR moves the addition and removal of the HUD to Login and Logout, respectively. In this way, AI bots will never have to worry about it, and player-controlled bots will be cleaned up properly.

The clean-up here does change existing behavior, but the existing behavior seems to have been written against two conflicting implementations. To be specific:

1. The code getting removed from `bot.dm` iterates through all of the diagnostic HUDs and adds them to the bot's display.
2. However, there's the field `/mob/living/simple_animal/bot/data_hud_type`, which, as a single value, suggests that each bot is only meant to have one diagnostic HUD.
3. Some bots manually add the hud specified in `data_hud_type`, without actually referencing it specifically.

My intuition is that this was a refactor that got put on hold and never followed through.

The result of the change is that whereas before, player-controlled bots would normally see all diagnostic HUDs, they now only see the one specified in `data_hud_type`. I think this is fine because 1) it makes sense that a sec bot wouldn't see the med HUD and vice versa, and 2) this is, relatively, incredibly niche behavior that only affects pAIs and people who possess simple bots, which is probably only admins.



## Why It's Good For The Game

GC failures bad.


## Testing


1. Checkout master.
2. Enable GC debugging options in `_compile_options.dm`.
3. Start server, spawn a Beepsky, and then delete it shortly after.
4. Repeat for Medibot, Honkbot, ED-209, Cleanbot, and Floorbot (Beepsky is representative of all secbots).
5. Observe GC failures logged as above.
6. Checkout this branch and repeat 2-4.
7. Observe there are no GC failures.

## Changelog
:cl:
fix: Free up refs causing GC failures for simple bots.
/:cl:
